river/Cursor.zig

```diff
@@ -747,26 +786,99 @@ fn processMotion(self: *Self, device: *wlr.InputDevice, time: u32, delta_x: f64,
+                    // means that the view has not maximum size, so we can
+                    // truncate without losing relevant data.
+                    const min_x = math.max(@intCast(i32, border_width), botright_x - @intCast(i32, @truncate(u16, constraints.max_width)));
+                    const max_x = botright_x - min_width;
+                    std.debug.assert(min_x <= max_x);
+                    box.x = math.clamp(box.x + @floatToInt(i32, delta_x), min_x, max_x);
+
+                    const min_y = math.max(@intCast(i32, border_width), botright_y - @intCast(i32, @truncate(u16, constraints.max_height)));
+                    const max_y = botright_y - min_height;
+                    std.debug.assert(min_y <= max_y);
+                    box.y = math.clamp(box.y + @floatToInt(i32, delta_y), min_y, max_y);
+                },
+                .botleft => {
+                    const min_x = math.max(@intCast(i32, border_width), botright_x - @intCast(i32, @truncate(u16, constraints.max_width)));
+                    const max_x = botright_x - min_width;
+                    std.debug.assert(min_x <= max_x);
+                    box.x = math.clamp(box.x + @floatToInt(i32, delta_x), min_x, max_x);
+
+                    const min_y = box.y + min_height;
+                    const max_y = box.y + @intCast(i32, @truncate(u16, constraints.max_height));
+                    std.debug.assert(min_y <= max_y);
+                    botright_y = math.clamp(botright_y + @floatToInt(i32, delta_y), min_y, max_y);
+                },
+                .topright => {
+                    const min_x = box.x + min_width;
+                    const max_x = box.x + @intCast(i32, @truncate(u16, constraints.max_width));
+                    std.debug.assert(min_x <= max_x);
+                    botright_x = math.clamp(botright_x + @floatToInt(i32, delta_x), min_x, max_x);
+
+                    const min_y = math.max(@intCast(i32, border_width), botright_y - @intCast(i32, @truncate(u16, constraints.max_height)));
+                    const max_y = botright_y - min_height;
+                    std.debug.assert(min_y <= max_y);
+                    box.y = math.clamp(box.y + @floatToInt(i32, delta_y), min_y, max_y);
+                },
+                .botright => {
+                    const min_x = box.x + min_width;
+                    const max_x = box.x + @intCast(i32, @truncate(u16, constraints.max_width));
+                    std.debug.assert(min_x <= max_x);
+                    botright_x = math.clamp(botright_x + @floatToInt(i32, delta_x), min_x, max_x);
+
+                    const min_y = box.y + min_height;
+                    const max_y = box.y + @intCast(i32, @truncate(u16, constraints.max_height));
+                    std.debug.assert(min_y <= max_y);
+                    botright_y = math.clamp(botright_y + @floatToInt(i32, delta_y), min_y, max_y);
+                },
+            }
```

Comment:
I think it should possible to avoid this level of code duplication. Perhaps something like the approach in tinywl would work:
https://github.com/swaywm/zig-wlroots/blob/c0703a0b59c2930b782e4811a2aedf4b6ab7fbae/tinywl/tinywl.zig#L275-L313